### DISCUSSION TITLE
'problem+json' error format

### DISCUSSION THREAD

**Discussion by janz93** [ruby-oauth/oauth2#547] [Ideas]
All our APIs are protected via oauth2 and we make use of this gem.

The [`OAuth2::Error class`](https://github.com/oauth-xx/oauth2/blob/ebe95171c3d34b9f8f4a4dfdca2e5094efc2deee/lib/oauth2/error.rb) is expecting the error response to always have an `error`, `error_description` parameter

Our APIs make use of the [`problem+json`](https://tools.ietf.org/html/rfc7807) as a structured format for error messages.

I was reading through the [oauth2 spec](https://tools.ietf.org/html/rfc6749#section-7.2) and if I understood the spec correctly it says that the resource error response format can look however you want it to be.

Maybe it would be great to support different error formats like `problem+json` 

PS. All thanks for the hard work that goes into this gem ❤️

**Comment by pboling** [comment]
Interesting idea.  I would have guessed, based solely on your description that "error response format" would be referring to the content and format of the **response body**, rather than the `CONTENT_TYPE`.  I would need to review the spec to understand what the requirements are for `CONTENT_TYPE`.

**Reply by janz93** [reply]
Ah my bad if it was a bit misleading 😅 

Sure, let me know if I can assist you somehow.

**Comment by pboling** [comment]
@janz93 I don't understand the issue you are describing.  According to https://datatracker.ietf.org/doc/html/rfc6749#section-5.2, when there is an error, the parameter `error` in the response is **required**.  `error_description` is **optional**.  What are you proposing that we change?  What specifically is the issue you are running into?

**Reply by pboling** [reply]
Can you create a PR with a spec showing the change you would like to see?

**Reply by janz93** [reply]
After looking at the Problem again I see that my initial idea is very complicated 
I think adding the problem+json to the JSON parser would be sufficient enough.
This way the `error.response.parsed` attribute can be used which makes the handling of the error fairly easy.

**Comment by pboling** [comment]
Support for `problem+json` was added by @janz93 in #553

**Reply by pboling** [reply]
@janz93 It has been released in version `2.0.0.rc1`.  Please try it out!
